oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,393 @@
+    ///
+    /// Returns an error if the sample is empty.
+    pub fn estimate(&self) -> Result<f64, QuantileError> {
+        // Return NaN if the sample is empty.
+        if self.is_empty() {
+            return Err(QuantileError::EmptyQuantile);
+        }
+
+        // Return the middle marker height if the sample size is at least 5.
+        if self.len() >= 5 {
+            return Ok(self.marker_hghts[2]);
+        }
+
+        let mut heights: [f64; 4] = [
+            self.marker_hghts[0],
+            self.marker_hghts[1],
+            self.marker_hghts[2],
+            self.marker_hghts[3],
+        ];
+
+        let len = self.len() as usize;
+        float_ord::sort(&mut heights[..len]);
+
+        let desired_index = (len as f64) * self._p() - 1.;
+        let mut index = desired_index.ceil();
+        if desired_index == index && index >= 0. {
+            let index = index.round_ties_even() as usize;
+            if index < len - 1 {
+                // `marker_hghts[index]` and `marker_hghts[index + 1]` are
+                // equally valid estimates,  by convention we take their average.
+                return Ok(0.5 * self.marker_hghts[index]
+                    + 0.5 * self.marker_hghts[index + 1]);
+            }
+        }
+        index = index.max(0.);
+        let mut index = index.round_ties_even() as usize;
+        index = min(index, len - 1);
+        Ok(self.marker_hghts[index])
+    }
+
+    /// Return the sample size.
+    pub fn len(&self) -> u64 {
+        self.marker_pos[4] as u64
+    }
+
+    /// Determine whether the sample is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Append a value/observation to the population and adjust the heights.
+    pub fn append<T>(&mut self, value: T) -> Result<(), QuantileError>
+    where
+        T: HistogramSupport,
+    {
+        if !value.is_finite() {
+            return Err(QuantileError::NonFiniteValue(format!("{:?}", value)));
+        }
+
+        // we've already checked that the value is finite.
+        let value = value.to_f64().unwrap();
+
+        // n[4] is the sample size.
+        if self.marker_pos[4] < 5 {
+            self.marker_hghts[self.marker_pos[4] as usize] = value;
+            self.marker_pos[4] += 1;
+            if self.marker_pos[4] == 5 {
+                float_ord::sort(&mut self.marker_hghts);
+            }
+            return Ok(());
+        }
+
+        // Find cell k.
+        let mut k: usize;
+        if value < self.marker_hghts[0] {
+            self.marker_hghts[0] = value;
+            k = 0;
+        } else {
+            k = 4;
+            for i in 1..5 {
+                if value < self.marker_hghts[i] {
+                    k = i;
+                    break;
+                }
+            }
+            if self.marker_hghts[4] < value {
+                self.marker_hghts[4] = value;
+            }
+        };
+
+        // Increment all positions greater than k.
```

Comment:
The comment seems to disagree with the code, since we're starting _at_ `k`. I think this is an artifact of the paper using 1-based indexing, possibly?